<img style="float: left; margin: 30px 15px 15px 15px;" src="https://pngimage.net/wp-content/uploads/2018/06/logo-iteso-png-5.png" width="300" height="500" /> 
    
    
### <font color='navy'> Simulación de procesos financieros. 

**Nombres:** Sean Nicolás González Vázquez y José Tonatiuh Navarro Silva.

**Fecha:** 17 de septiembre del 2020.

**Expediente** : 720083 y 723399
**Profesor:** Oscar David Jaramillo Zuluaga.

# Tarea 3

### El link donde se evidencia el trabajo es el siguiente: [Perfil GitHub Nicolas Gonzalez](https://github.com/nicoglez)

## Enunciado de tarea 

Usando compresión de listas o funciones map(sino recuerda como funciona observar el siguiente enlace https://www.pythonforbeginners.com/lists/list-comprehensions-in-python/), resolver los siguientes ejercicios:  

#### **Ejercicio 1**

1. Resolver la siguiente ecuación recursiva usando funciones como se vió en clase
$$ D_{n}=(n-1) D_{n-1}+(n-1) D_{n-2} \quad n\ge 3$$
con $D_1=0$ y $D_2 = 1$

#### **Ejercicio 2**
2. Count the number of spaces in the following string `variable =  relaciónn  requiere,  para obtener  el  valor  de un  cierto  término,  el conocimiento de los dos anteriores`.

#### **Ejercicio 3**
3. Remove all of the vowels in a string [make a list of the non-vowels].

#### **Ejercicio 4**
4. Find all of the words in a string that are less than 4 letters.

#### **Ejercicio 5**
5. Use a dictionary comprehension to count the length of each word in a sentence.

#### **Ejercicio 6**
6. Use a nested list comprehension to find all of the numbers from 1-1000 that are divisible by any single digit besides 1 (2-9). 

In [2]:
# importar librerias
import numpy as np

### Ejercicio 1: 

1. Resolver la siguiente ecuación recursiva usando funciones como se vió en clase
$$ D_{n}=(n-1) D_{n-1}+(n-1) D_{n-2} \quad n\ge 3$$
con $D_1=0$ y $D_2 = 1$



In [11]:
# Solucion Nicolas Glez
def r(n: 'Numeros de elementos (-1) a obtener'):
    '''
    Esta funcion sirve para encontrar la resolucion de la ecuacion recursiva:
    $$ D_{n}=(n-1) D_{n-1}+(n-1) D_{n-2} para n > 3 con D_1 =0  y D_2 = 1$
    '''
    D = np.zeros(n)
    D[0], D[1] = 0,1
    
    def DD(i):
        nonlocal D
        D[i-1] = (i-1) * (D[i-2]) + (i-1) * (D[i-3])
    [DD(x) for x in range(3, n+1)]
    
    return D

r(6)

array([  0.,   1.,   2.,   9.,  44., 265.])

In [4]:
# Solucion Jose Tonatiuh

### **Ejercicio 2:**
2. Count the number of spaces in the following string `variable =  relaciónn  requiere,  para obtener  el  valor  de un  cierto  término,  el conocimiento de los dos anteriores`.


In [5]:
# Solucion Nicolas Glez
variable =  'relaciónn  requiere,  para obtener  el  valor  de un  cierto  término,  el conocimiento de los dos anteriores'

count_spaces = np.sum([len(words) for words in variable if words == ' '])
count_spaces

23

In [6]:
# Solucion Jose Tonatiuh

### **Ejercicio 3**
3. Remove all of the vowels in a string [make a list of the non-vowels].



In [8]:
# Solucion Nicolas Glez
variable =  'relaciónn  requiere,  para obtener  el  valor  de un  cierto  término,  el conocimiento de los dos anteriores'
filt = [''.join([word for word in variable if (word != 'a' and word != 'e' and word != 'i' and word != 'o' and word != 'u' and
                                               word != 'á' and word != 'é' and word != 'í' and word != 'ó' and word != 'ú')])]
print(filt)

['rlcnn  rqr,  pr btnr  l  vlr  d n  crt  trmn,  l cncmnt d ls ds ntrrs']


In [ ]:
# Solucion Jose Tonatiuh

### **Ejercicio 4**
4. Find all of the words in a string that are less than 4 letters.


In [14]:
# Solucion Nicolas Glez
variable =  'relaciónn  requiere,  para obtener  el  valor  de un  cierto  término,  el conocimiento de los dos anteriores'
s4 = [less for less in variable.split() if len(less)<4]
s4

['el', 'de', 'un', 'el', 'de', 'los', 'dos']

In [ ]:
# Solucion Jose Tonatiuh

### **Ejercicio 5**
5. Use a dictionary comprehension to count the length of each word in a sentence.
 

In [14]:
# Solucion Nicolas Glez
variable =  'relaciónn  requiere,  para obtener  el  valor  de un  cierto  término,  el conocimiento de los dos anteriores'
s5 = {word: len(word) for word in variable.split()}
s5 

{'relaciónn': 9,
 'requiere,': 9,
 'para': 4,
 'obtener': 7,
 'el': 2,
 'valor': 5,
 'de': 2,
 'un': 2,
 'cierto': 6,
 'término,': 8,
 'conocimiento': 12,
 'los': 3,
 'dos': 3,
 'anteriores': 10}

In [ ]:
# Solucion Jose Tonatiuh

### **Ejercicio 6**
6. Use a nested list comprehension to find all of the numbers from 1-1000 that are divisible by any single digit besides 1 (2-9).

In [15]:
# Solucion Nicolas Glez
l = [numbers for numbers in np.arange(1,1001) if any (numbers % digits == 0 for digits in np.arange(2,10))]

print(l)

[2, 3, 4, 5, 6, 7, 8, 9, 10, 12, 14, 15, 16, 18, 20, 21, 22, 24, 25, 26, 27, 28, 30, 32, 33, 34, 35, 36, 38, 39, 40, 42, 44, 45, 46, 48, 49, 50, 51, 52, 54, 55, 56, 57, 58, 60, 62, 63, 64, 65, 66, 68, 69, 70, 72, 74, 75, 76, 77, 78, 80, 81, 82, 84, 85, 86, 87, 88, 90, 91, 92, 93, 94, 95, 96, 98, 99, 100, 102, 104, 105, 106, 108, 110, 111, 112, 114, 115, 116, 117, 118, 119, 120, 122, 123, 124, 125, 126, 128, 129, 130, 132, 133, 134, 135, 136, 138, 140, 141, 142, 144, 145, 146, 147, 148, 150, 152, 153, 154, 155, 156, 158, 159, 160, 161, 162, 164, 165, 166, 168, 170, 171, 172, 174, 175, 176, 177, 178, 180, 182, 183, 184, 185, 186, 188, 189, 190, 192, 194, 195, 196, 198, 200, 201, 202, 203, 204, 205, 206, 207, 208, 210, 212, 213, 214, 215, 216, 217, 218, 219, 220, 222, 224, 225, 226, 228, 230, 231, 232, 234, 235, 236, 237, 238, 240, 242, 243, 244, 245, 246, 248, 249, 250, 252, 254, 255, 256, 258, 259, 260, 261, 262, 264, 265, 266, 267, 268, 270, 272, 273, 274, 275, 276, 278, 279, 280, 282,

In [ ]:
# Solucion Jose Tonatiuh

### Respuesta planteada en el ejercicio usando (markdown)

La respuesta del **<font color= blue>Ejercicio 1<font color= blue>** es, que los primeros seis terminos de la sucesion recursiva:

$$ D_{n}=(n-1) D_{n-1}+(n-1) D_{n-2} \quad n\ge 3$$
con $D_1=0$ y $D_2 = 1$

Son los siguientes:

```array([  0.,   1.,   2.,   9.,  44., 265.])```



La respuesta del **<font color= blue>Ejercicio 2<font color= blue>** es que el numero de espacios en el string:

`variable =  relaciónn  requiere,  para obtener  el  valor  de un  cierto  término,  el conocimiento de los dos anteriores`

Son **```23```**

La respuesta del **<font color= blue>Ejercicio 3<font color= blue>** es que si quitamos las vocales del string:

`variable =  'relaciónn  requiere,  para obtener  el  valor  de un  cierto  término,  el conocimiento de los dos anteriores'`

Obtenemos:  

```['rlcnn  rqr,  pr btnr  l  vlr  d n  crt  trmn,  l cncmnt d ls ds ntrrs']```

La respuesta del **<font color= blue>Ejercicio 4<font color= blue>** es que si filtramos todas las palabras con longitud menor a 4 del string:

`variable =  'relaciónn  requiere,  para obtener  el  valor  de un  cierto  término,  el conocimiento de los dos anteriores'`

Obtenemos

```['el', 'de', 'un', 'el', 'de', 'los', 'dos']```

La respuesta del **<font color= blue>Ejercicio 5<font color= blue>** es que si usamos compresion de diccionarios para hacer un codigo que cuente la longitud de cada palabra, obtenemos que del string:

`variable =  'relaciónn  requiere,  para obtener  el  valor  de un  cierto  término,  el conocimiento de los dos anteriores'`


La longitud de cada una de las palabras es:

`{'relaciónn': 9,
 'requiere,': 9,
 'para': 4,
 'obtener': 7,
 'el': 2,
 'valor': 5,
 'de': 2,
 'un': 2,
 'cierto': 6,
 'término,': 8,
 'conocimiento': 12,
 'los': 3,
 'dos': 3,
 'anteriores': 10}`

La respuesta del **<font color= blue>Ejercicio 6<font color= blue>** es que todos los digitos del 1 al 1000 divisibles entre un numero de un digito excepto 1 (2 al 9), son los siguientes.

`[2, 3, 4, 5, 6, 7, 8, 9, 10, 12, 14, 15, 16, 18, 20, 21, 22, 24, 25, 26, 27, 28, 30, 32, 33, 34, 35, 36, 38, 39, 40, 42, 44, 45, 46, 48, 49, 50, 51, 52, 54, 55, 56, 57, 58, 60, 62, 63, 64, 65, 66, 68, 69, 70, 72, 74, 75, 76, 77, 78, 80, 81, 82, 84, 85, 86, 87, 88, 90, 91, 92, 93, 94, 95, 96, 98, 99, 100, 102, 104, 105, 106, 108, 110, 111, 112, 114, 115, 116, 117, 118, 119, 120, 122, 123, 124, 125, 126, 128, 129, 130, 132, 133, 134, 135, 136, 138, 140, 141, 142, 144, 145, 146, 147, 148, 150, 152, 153, 154, 155, 156, 158, 159, 160, 161, 162, 164, 165, 166, 168, 170, 171, 172, 174, 175, 176, 177, 178, 180, 182, 183, 184, 185, 186, 188, 189, 190, 192, 194, 195, 196, 198, 200, 201, 202, 203, 204, 205, 206, 207, 208, 210, 212, 213, 214, 215, 216, 217, 218, 219, 220, 222, 224, 225, 226, 228, 230, 231, 232, 234, 235, 236, 237, 238, 240, 242, 243, 244, 245, 246, 248, 249, 250, 252, 254, 255, 256, 258, 259, 260, 261, 262, 264, 265, 266, 267, 268, 270, 272, 273, 274, 275, 276, 278, 279, 280, 282, 284, 285, 286, 287, 288, 290, 291, 292, 294, 295, 296, 297, 298, 300, 301, 302, 303, 304, 305, 306, 308, 309, 310, 312, 314, 315, 316, 318, 320, 321, 322, 324, 325, 326, 327, 328, 329, 330, 332, 333, 334, 335, 336, 338, 339, 340, 342, 343, 344, 345, 346, 348, 350, 351, 352, 354, 355, 356, 357, 358, 360, 362, 363, 364, 365, 366, 368, 369, 370, 371, 372, 374, 375, 376, 378, 380, 381, 382, 384, 385, 386, 387, 388, 390, 392, 393, 394, 395, 396, 398, 399, 400, 402, 404, 405, 406, 408, 410, 411, 412, 413, 414, 415, 416, 417, 418, 420, 422, 423, 424, 425, 426, 427, 428, 429, 430, 432, 434, 435, 436, 438, 440, 441, 442, 444, 445, 446, 447, 448, 450, 452, 453, 454, 455, 456, 458, 459, 460, 462, 464, 465, 466, 468, 469, 470, 471, 472, 474, 475, 476, 477, 478, 480, 482, 483, 484, 485, 486, 488, 489, 490, 492, 494, 495, 496, 497, 498, 500, 501, 502, 504, 505, 506, 507, 508, 510, 511, 512, 513, 514, 515, 516, 518, 519, 520, 522, 524, 525, 526, 528, 530, 531, 532, 534, 535, 536, 537, 538, 539, 540, 542, 543, 544, 545, 546, 548, 549, 550, 552, 553, 554, 555, 556, 558, 560, 561, 562, 564, 565, 566, 567, 568, 570, 572, 573, 574, 575, 576, 578, 579, 580, 581, 582, 584, 585, 586, 588, 590, 591, 592, 594, 595, 596, 597, 598, 600, 602, 603, 604, 605, 606, 608, 609, 610, 612, 614, 615, 616, 618, 620, 621, 622, 623, 624, 625, 626, 627, 628, 630, 632, 633, 634, 635, 636, 637, 638, 639, 640, 642, 644, 645, 646, 648, 650, 651, 652, 654, 655, 656, 657, 658, 660, 662, 663, 664, 665, 666, 668, 669, 670, 672, 674, 675, 676, 678, 679, 680, 681, 682, 684, 685, 686, 687, 688, 690, 692, 693, 694, 695, 696, 698, 699, 700, 702, 704, 705, 706, 707, 708, 710, 711, 712, 714, 715, 716, 717, 718, 720, 721, 722, 723, 724, 725, 726, 728, 729, 730, 732, 734, 735, 736, 738, 740, 741, 742, 744, 745, 746, 747, 748, 749, 750, 752, 753, 754, 755, 756, 758, 759, 760, 762, 763, 764, 765, 766, 768, 770, 771, 772, 774, 775, 776, 777, 778, 780, 782, 783, 784, 785, 786, 788, 789, 790, 791, 792, 794, 795, 796, 798, 800, 801, 802, 804, 805, 806, 807, 808, 810, 812, 813, 814, 815, 816, 818, 819, 820, 822, 824, 825, 826, 828, 830, 831, 832, 833, 834, 835, 836, 837, 838, 840, 842, 843, 844, 845, 846, 847, 848, 849, 850, 852, 854, 855, 856, 858, 860, 861, 862, 864, 865, 866, 867, 868, 870, 872, 873, 874, 875, 876, 878, 879, 880, 882, 884, 885, 886, 888, 889, 890, 891, 892, 894, 895, 896, 897, 898, 900, 902, 903, 904, 905, 906, 908, 909, 910, 912, 914, 915, 916, 917, 918, 920, 921, 922, 924, 925, 926, 927, 928, 930, 931, 932, 933, 934, 935, 936, 938, 939, 940, 942, 944, 945, 946, 948, 950, 951, 952, 954, 955, 956, 957, 958, 959, 960, 962, 963, 964, 965, 966, 968, 969, 970, 972, 973, 974, 975, 976, 978, 980, 981, 982, 984, 985, 986, 987, 988, 990, 992, 993, 994, 995, 996, 998, 999, 1000]`